### PR TITLE
[EasyCore] Doctrine helpers

### PR DESCRIPTION
**Pull request by vlukiashka**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

https://eonx.atlassian.net/browse/PLU-3244

### CODE REVIEW

File: packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+            $expectedEntityInsertions,
+            $this->getPrivatePropertyValue($deferredEntityEventDispatcher, 'entityInsertions')
+        );
+        self::assertSame(
+            $expectedEntityUpdates,
+            $this->getPrivatePropertyValue($deferredEntityEventDispatcher, 'entityUpdates')
+        );
+    }
+
+    public function testDeferInsertionsSucceeds(): void
```
Comment (akl-95) [inline packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php:141]: let's cover case with enabled = false

File: packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+use EonX\EasyCore\Doctrine\Events\EntityUpdatedEvent;
+use EonX\EasyCore\Interfaces\DatabaseEntityInterface;
+use EonX\EasyCore\Tests\AbstractTestCase;
+use Prophecy\Argument;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * @covers \EonX\EasyCore\Doctrine\Dispatchers\DeferredEntityEventDispatcher
+ */
+final class DeferredEntityEventDispatcherTest extends AbstractTestCase
```
Comment (akl-95) [inline packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php:18]: Pls cover the disable and enable methods.

**Review by akl-95 — APPROVED** [review]

**Review by akl-95 — CHANGES_REQUESTED** [review]
Please provide fixes according to my comments.

File: packages/EasyCore/src/Doctrine/EntityManagers/EntityManagerDecorator.php [hunk 3]
```diff
@@ -0,0 +1,94 @@
+<?php
+
+declare(strict_types=1);
+
+namespace EonX\EasyCore\Doctrine\EntityManagers;
```
[the remark this answers is not in the document]
Comment (vlukiashka) [reply packages/EasyCore/src/Doctrine/EntityManagers/EntityManagerDecorator.php]: Done

File: packages/EasyCore/tests/Doctrine/EntityManagers/EntityManagerDecoratorTest.php [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+use EonX\EasyCore\Tests\AbstractTestCase;
+use EonX\EasyErrorHandler\Interfaces\ErrorHandlerInterface;
+use Exception;
+use InvalidArgumentException;
+use stdClass;
+
+/**
+ * @covers \EonX\EasyCore\Doctrine\EntityManagers\EntityManagerDecorator
+ */
+final class EntityManagerDecoratorTest extends AbstractTestCase
```
[the remark this answers is not in the document]
Comment (vlukiashka) [reply packages/EasyCore/tests/Doctrine/EntityManagers/EntityManagerDecoratorTest.php:22]: Done

File: packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php (hunk 1 above)
Comment (vlukiashka) [reply packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php:141]: Done

File: packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php (hunk 2 above)
Comment (vlukiashka) [reply packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php:18]: Done

File: packages/EasyCore/tests/Doctrine/EntityManagers/EntityManagerDecoratorTest.php [hunk 5]
```diff
@@ -0,0 +1,386 @@
+<?php
+
+declare(strict_types=1);
+
+namespace EonX\EasyCore\Tests\Doctrine\ORM\Decorators;
```
Comment (akl-95) [inline packages/EasyCore/tests/Doctrine/EntityManagers/EntityManagerDecoratorTest.php:5]: Pls move this file to the corresponding folder.

File: packages/EasyCore/tests/Doctrine/EntityManagers/EntityManagerDecoratorTest.php [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+        $entityManagerDecorator = new EntityManagerDecorator(
+            $eventDispatcherReveal,
+            $errorHandlerReveal,
+            $entityManagerReveal
+        );
+
+        $entityManagerDecorator->rollback();
+
+        $entityManager->rollback()
+            ->shouldNotBeCalled();
```
Comment (akl-95) [inline packages/EasyCore/tests/Doctrine/EntityManagers/EntityManagerDecoratorTest.php:151]: Pls use shouldNotHaveBeenCalled instead.

File: packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+        self::assertSame([], $this->getPrivatePropertyValue($deferredEntityEventDispatcher, 'entityUpdates'));
+    }
+
+    public function testEnableSucceeds(): void
+    {
+        /** @var \Symfony\Contracts\EventDispatcher\EventDispatcherInterface $eventDispatcherReveal */
+        $eventDispatcherReveal = $this->prophesize(EventDispatcherInterface::class)->reveal();
+        $deferredEntityEventDispatcher = new DeferredEntityEventDispatcher($eventDispatcherReveal);
+
+        $deferredEntityEventDispatcher->disable();
```
Comment (akl-95) [inline packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php:292]: Let's move this line to the arrange block.

File: packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+        $deferredEntityEventDispatcher->enable();
+
+        self::assertTrue($this->getPrivatePropertyValue($deferredEntityEventDispatcher, 'enabled'));
+    }
+
+    public function testEnableSucceedsDefaultValue(): void
```
Comment (akl-95) [inline packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php:298]: Let's name it like testEnabledSucceedsAndContainsTrueByDefault

File: packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+    {
+        $entityA = $this->prophesize(DatabaseEntityInterface::class)->reveal();
+        $entityB = $this->prophesize(DatabaseEntityInterface::class)->reveal();
+        /** @var \EonX\EasyCore\Interfaces\DatabaseEntityInterface[] $entityAB */
+        $entityAB = [$entityA, $entityB];
+        /** @var \Symfony\Contracts\EventDispatcher\EventDispatcherInterface $eventDispatcherReveal */
+        $eventDispatcherReveal = $this->prophesize(EventDispatcherInterface::class)->reveal();
+        $deferredEntityEventDispatcher = new DeferredEntityEventDispatcher($eventDispatcherReveal);
+
+        $deferredEntityEventDispatcher->disable();
```
Comment (akl-95) [inline packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php:176]: Pls move it to the arrange block.

File: packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
```
Comment (akl-95) [inline packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php:220]: The same here.

**Review by akl-95 — CHANGES_REQUESTED** [review]
Please provide fixes according to my comments.

File: packages/EasyCore/tests/Doctrine/EntityManagers/EntityManagerDecoratorTest.php (hunk 5 above)
Comment (vlukiashka) [reply packages/EasyCore/tests/Doctrine/EntityManagers/EntityManagerDecoratorTest.php]: Done

File: packages/EasyCore/tests/Doctrine/EntityManagers/EntityManagerDecoratorTest.php (hunk 6 above)
Comment (vlukiashka) [reply packages/EasyCore/tests/Doctrine/EntityManagers/EntityManagerDecoratorTest.php]: Done

File: packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php (hunk 7 above)
Comment (vlukiashka) [reply packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php:291]: Done

File: packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php (hunk 8 above)
Comment (vlukiashka) [reply packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php]: Done

File: packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php (hunk 10 above)
Comment (vlukiashka) [reply packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php:219]: Done

**Review by akl-95 — DISMISSED** [review]

File: packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php [hunk 11]
```diff
@@ -0,0 +1,119 @@
+<?php
+
+declare(strict_types=1);
+
+namespace EonX\EasyCore\Doctrine\Dispatchers;
+
+use EonX\EasyCore\Doctrine\Events\EntityCreatedEvent;
+use EonX\EasyCore\Doctrine\Events\EntityUpdatedEvent;
+use EonX\EasyCore\Interfaces\DatabaseEntityInterface;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
```
Comment (itorgov) [inline packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php:10]: Please replace it with `\Psr\EventDispatcher\EventDispatcherInterface`. We will use `DeferredEntityEventDispatcher` in Lumen projects too (Lumen doesn't have the `Symfony\Contracts\EventDispatcher\EventDispatcherInterface` interface and uses `\Psr\EventDispatcher\EventDispatcherInterface` instead).

**Review by itorgov — CHANGES_REQUESTED** [review]

File: packages/EasyCore/src/Interfaces/DatabaseEntityInterface.php [hunk 12]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+declare(strict_types=1);
+
+namespace EonX\EasyCore\Interfaces;
+
+interface DatabaseEntityInterface
+{
+    public function getId(): string;
```
Comment (itorgov) [inline packages/EasyCore/src/Interfaces/DatabaseEntityInterface.php:9]: Let's left this interface empty. In some projects (like Points) we have a different signature of this method.

**Review by itorgov — CHANGES_REQUESTED** [review]

File: packages/EasyCore/src/Doctrine/Events/EntityCreatedEvent.php [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+use EonX\EasyCore\Interfaces\DatabaseEntityInterface;
+use Symfony\Contracts\EventDispatcher\Event;
+
+final class EntityCreatedEvent extends Event
+{
+    /**
+     * @var string
+     */
+    public const NAME = 'entity.created';
```
Comment (roman-eonx) [inline packages/EasyCore/src/Doctrine/Events/EntityCreatedEvent.php:15]: We don't need this const anymore, right?

**Review by roman-eonx — CHANGES_REQUESTED** [review]

File: packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php (hunk 11 above)
Comment (vlukiashka) [reply packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php]: Done

File: packages/EasyCore/src/Interfaces/DatabaseEntityInterface.php (hunk 12 above)
Comment (vlukiashka) [reply packages/EasyCore/src/Interfaces/DatabaseEntityInterface.php]: Done

File: packages/EasyCore/src/Doctrine/Events/EntityCreatedEvent.php (hunk 13 above)
Comment (vlukiashka) [reply packages/EasyCore/src/Doctrine/Events/EntityCreatedEvent.php]: Done

**Review by itorgov — APPROVED** [review]

**Review by roman-eonx — DISMISSED** [review]

File: packages/EasyCore/tests/Doctrine/Subscribers/EntityEventSubscriberTest.php [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+        $entityEventSubscriber = new EntityEventSubscriber(
+            $deferredEntityEventDispatcherReveal,
+            [\get_class($newEntity), \get_class($existedEntity)]
+        );
+        /** @var \Doctrine\ORM\Event\OnFlushEventArgs $eventArgsReveal */
+        $eventArgsReveal = $eventArgs->reveal();
+
+        $entityEventSubscriber->onFlush($eventArgsReveal);
+
+        $eventArgs->getEntityManager()
```
Comment (itorgov) [inline packages/EasyCore/tests/Doctrine/Subscribers/EntityEventSubscriberTest.php:70]: Why is it written in two lines? I found many cases of this in this PR.

**Review by itorgov — CHANGES_REQUESTED** [review]

**Review by itorgov — DISMISSED** [review]

File: packages/EasyCore/src/Doctrine/Events/EntityUpdatedEvent.php [hunk 15]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+declare(strict_types=1);
+
+namespace EonX\EasyCore\Doctrine\Events;
+
+use EonX\EasyCore\Interfaces\DatabaseEntityInterface;
+use Symfony\Contracts\EventDispatcher\Event;
```
Comment (natepage) [inline packages/EasyCore/src/Doctrine/Events/EntityUpdatedEvent.php:8]: Do we really need to extend the symfony event class?
The idea of this is to work with doctrine regardless of the framework it is used with.

File: packages/EasyCore/src/Interfaces/DatabaseEntityInterface.php [hunk 16]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+declare(strict_types=1);
+
+namespace EonX\EasyCore\Interfaces;
+
+interface DatabaseEntityInterface
```
Comment (natepage) [inline packages/EasyCore/src/Interfaces/DatabaseEntityInterface.php:7]: Does it mean that every project will have to make their entities implement this interface for the logic to work?

Why is that needed?

File: packages/EasyCore/src/Doctrine/Subscribers/EntityEventSubscriberInterface.php [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace EonX\EasyCore\Doctrine\Subscribers;
+
+use Doctrine\Common\EventSubscriber;
+use Doctrine\ORM\Event\OnFlushEventArgs;
+use Doctrine\ORM\Event\PostFlushEventArgs;
+
+interface EntityEventSubscriberInterface extends EventSubscriber
+{
+    public function getSubscribedEvents(): array;
```
Comment (natepage) [inline packages/EasyCore/src/Doctrine/Subscribers/EntityEventSubscriberInterface.php:13]: PHPDoc are required for iterable arguments/return types.

File: packages/EasyCore/src/Doctrine/Subscribers/EntityEventSubscriber.php [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+    /**
+     * @param string[] $entities
+     */
+    public function __construct(DeferredEntityEventDispatcherInterface $eventDispatcher, array $entities)
+    {
+        $this->eventDispatcher = $eventDispatcher;
+        $this->entities = $entities;
+    }
+
+    public function getSubscribedEvents(): array
```
Comment (natepage) [inline packages/EasyCore/src/Doctrine/Subscribers/EntityEventSubscriber.php:34]: PHPDoc here too please.

File: packages/EasyCore/src/Doctrine/ORM/Decorators/EntityManagerDecorator.php [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+
+        $this->beginTransaction();
+
+        try {
+            $return = $func($this);
+
+            $this->flush();
+            $this->commit();
+
+            return $return ?? true;
```
Comment (natepage) [inline packages/EasyCore/src/Doctrine/ORM/Decorators/EntityManagerDecorator.php:81]: Do we need to default to true?
What if some app wants not to return anything from `$func` on purpose?

**Review by natepage — CHANGES_REQUESTED** [review]

File: packages/EasyCore/src/Interfaces/DatabaseEntityInterface.php (hunk 16 above)
Comment (vlukiashka) [reply packages/EasyCore/src/Interfaces/DatabaseEntityInterface.php:7]: We are filtering entities by this interface in the current implementation.
So, looks like we need it.

File: packages/EasyCore/src/Doctrine/Events/EntityUpdatedEvent.php (hunk 15 above)
Comment (vlukiashka) [reply packages/EasyCore/src/Doctrine/Events/EntityUpdatedEvent.php]: Not needed. Removed.

File: packages/EasyCore/src/Doctrine/Subscribers/EntityEventSubscriberInterface.php (hunk 17 above)
Comment (vlukiashka) [reply packages/EasyCore/src/Doctrine/Subscribers/EntityEventSubscriberInterface.php:16]: Done

File: packages/EasyCore/src/Doctrine/Subscribers/EntityEventSubscriber.php (hunk 18 above)
Comment (vlukiashka) [reply packages/EasyCore/src/Doctrine/Subscribers/EntityEventSubscriber.php:37]: Done

File: packages/EasyCore/src/Doctrine/ORM/Decorators/EntityManagerDecorator.php (hunk 19 above)
Comment (vlukiashka) [reply packages/EasyCore/src/Doctrine/ORM/Decorators/EntityManagerDecorator.php]: Not needed. Fixed.

File: packages/EasyCore/src/Interfaces/DatabaseEntityInterface.php (hunk 16 above)
Comment (natepage) [reply packages/EasyCore/src/Interfaces/DatabaseEntityInterface.php:7]: Is it a requirement from doctrine for a class to implement a specific interface?
No, so why do we need an interface?

Comment (vlukiashka) [reply packages/EasyCore/src/Interfaces/DatabaseEntityInterface.php]: We don't need it. Removed

**Review by natepage — APPROVED** [review]
